### PR TITLE
8273297: AES/GCM non-AVX512+VAES CPUs suffer after 8267125

### PR DESCRIPTION
**Pull request by smita-kamath**
Performance dropped up to 10% for 1k data after 8267125 for CPUs that do not support the new intrinsic. Tests run were crypto.full.AESGCMBench and crypto.full.AESGCMByteBuffer from the jmh micro benchmarks.

The problem is each instance of GHASH allocates 96 extra longs for the AVX512+VAES intrinsic regardless if the intrinsic is used. This extra table space should be allocated differently so that non-supporting CPUs do not suffer this penalty. This issue also affects non-Intel CPUs too.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273297](https://bugs.openjdk.java.net/browse/JDK-8273297): AES/GCM non-AVX512+VAES CPUs suffer after 8267125

### Reviewers
 * [Anthony Scarpino](https://openjdk.java.net/census#ascarpino) (@ascarpino - **Reviewer**) ⚠️ Review applies to 7ea464aece8eb9ca63606e850875825ab21ea9b7
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to 19b0d547d1b05f9e6c133b8fa7876b7283a21095
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5402/head:pull/5402` \
`$ git checkout pull/5402`

Update a local copy of the PR: \
`$ git checkout pull/5402` \
`$ git pull https://git.openjdk.java.net/jdk pull/5402/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5402`

View PR using the GUI difftool: \
`$ git pr show -t 5402`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5402.diff">https://git.openjdk.java.net/jdk/pull/5402.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/library_call.cpp [hunk 1]
```diff
@@ -6789,11 +6789,20 @@ bool LibraryCallKit::inline_galoisCounterMode_AESCrypt() {
   Node* state_start = array_element_address(state, intcon(0), T_LONG);
   Node* subkeyHtbl_start = array_element_address(subkeyHtbl, intcon(0), T_LONG);
 
+  ciKlass* klass = ciTypeArrayKlass::make(T_LONG);
+  Node* klass_node = makecon(TypeKlassPtr::make(klass));
+
+  Node* avx512_subkeyHtbl = new_array(klass_node, intcon(96), 0);
+  if (avx512_subkeyHtbl == NULL) return false;
```
Comment (theRealAph) [inline src/hotspot/share/opto/library_call.cpp:6796]: This looks very Intel-specific, but it's in generic code. Please make this constant 96 a symbol and push it into a header file in the x86 back end.

**Comment by theRealAph** [comment]
It seems to me there's a serious problem here. When you execute the galoisCounterMode_AESCrypt() intrinsic, I don't think there's a limit on the number of blocks to be encrypted. With the older intrinsic things are not so very bad because the incoming data is split into 6 segments. But if we use this intrinsic, there is no safepoint check in the inner loop, which can lead to a long time to safepoint, and this causes stalls on the other threads.
If you split the incoming data into blocks of about a megabyte you'd lose no measurable performance but you'd dramatically improve the performance of everything else, especially with a concurrent GC.

File: src/hotspot/share/opto/library_call.cpp (hunk 1 above)
Comment (smita-kamath) [reply src/hotspot/share/opto/library_call.cpp:6796]: I'll modify the code. Thanks.

**Comment by smita-kamath** [comment]
@theRealAph Thank you for the comment above. I will look into this issue.

**Review by ascarpino — APPROVED** [review]
I approve the jdk changes.   You'll need a hotspot reviewer to approve the other changes

**Comment by smita-kamath** [comment]
@theRealAph I have implemented changes as per your suggestions. Could you review the changes and let me know your thoughts?

File: src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java [hunk 2]
```diff
@@ -570,6 +572,28 @@ static void increment32(byte[] value) {
...
+                                    int ctOfs, byte[] out, int outOfs,
+                                    GCTR gctr, GHASH ghash) {
+
+        int len = 0;
+        if (inLen > SPLIT_LEN) {
+            while (inLen >= SPLIT_LEN) {
+                int partlen = implGCMCrypt0(in, inOfs+len, SPLIT_LEN, ct,
+                    ctOfs+len, out, outOfs+len, gctr, ghash);
+                len+= partlen;
+                inLen-= len;
```
Comment (sviswa7) [inline src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java:588]: This should be inLen -= partlen;

Comment (smita-kamath) [reply src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java]: Done. Thank you for pointing this out.

**Comment by sviswa7** [comment]
The patch looks good to me.

**Review by sviswa7 — APPROVED** [review]

**Comment by smita-kamath** [comment]
@ascarpino Could you please run tier 1-3 tests? We have two reviewers for the patch.

**Comment by smita-kamath** [comment]
@ascarpino I dont see your comments on this PR. Could you please post them again?

**Comment by smita-kamath** [comment]
@ascarpino I've modified the code style. Thank you.

**Comment by sviswa7** [comment]
Hotspot changes look good.

**Comment by smita-kamath** [comment]
@ascarpino Is it okay to integrate this patch?

**Comment by ascarpino** [comment]
I think it's ready to integrate

**Review by theRealAph — APPROVED** [review]

**Pull request closed without merging**